include/oneapi/dpl/internal/distributed_ranges_impl/shp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,260 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <optional>
+
+#include <sycl/sycl.hpp>
+
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/numeric>
+
+#include <oneapi/dpl/async>
+
+#include <oneapi/dpl/internal/distributed_ranges_impl/concepts/concepts.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/detail/onedpl_direct_iterator.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/algorithms/execution_policy.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/allocators.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/detail.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/init.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/vector.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/shp/views/views.hpp>
+
+namespace oneapi::dpl::experimental::dr::shp
+{
+
+template <typename ExecutionPolicy, distributed_contiguous_range R, distributed_contiguous_range O, typename BinaryOp,
+          typename U = rng::range_value_t<R>>
+void
+inclusive_scan_impl_(ExecutionPolicy&& policy, R&& r, O&& o, BinaryOp binary_op, std::optional<U> init = {})
+{
+    using T = rng::range_value_t<O>;
+
+    static_assert(std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, distributed_device_policy>);
+
+    auto zipped_view = views::zip(r, o);
+    auto zipped_segments = zipped_view.zipped_segments();
+
+    if constexpr (std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, distributed_device_policy>)
+    {
+
+        std::vector<sycl::event> events;
+
+        auto root = devices()[0];
+        device_allocator<T> allocator(context(), root);
+        vector<T, device_allocator<T>> partial_sums(std::size_t(zipped_segments.size()), allocator);
+
+        std::size_t segment_id = 0;
+        for (auto&& segs : zipped_segments)
+        {
+            auto&& [in_segment, out_segment] = segs;
+
+            auto&& q = __detail::queue(ranges::rank(in_segment));
+            auto&& local_policy = __detail::dpl_policy(ranges::rank(in_segment));
+
+            auto dist = rng::distance(in_segment);
+            assert(dist > 0);
+
+            auto first = rng::begin(in_segment);
+            auto last = rng::end(in_segment);
+            auto d_first = rng::begin(out_segment);
+
+            sycl::event event;
```

Comment:
This might be problematic in the future. Here, we don't keep the future-like object alive, as returned by `scan_async` but implicitly convert it to a `sycl::event`. This can represent an implicit synchronization point. Shouldn't be an issue if `sycl::buffer` are used for temporary memory allocations, like in our current implementation of prefix scan. But that might change. We should at least verify the behavior is as expected, I'm also not sure we have any tests that reflect this.

Reply:
Yes, in fact we have active plans to move to `__result_and_scratch_storage` for temporary storage, which may use USM memory and a shared_ptr kept alive by the future value.  This is in our prototype branch for two_pass_scan which we are targeting for 2025.0 as well.  Reduce already does this.
If I understand correctly, the future leaving scope could free the temporary storage and return early before the kernel returns and this would cause problems.  
I don't know of other plans to convert to USM memory as temporary allocations, but I think we should probably assume that we need to keep the future alive until it is waited upon.

Reply:
Thanks for pointing this out.  It seems like the simplest fix would be to simply add a new vector that stores all the oneDPL futures, keeping them alive while leaving the other asynchronous logic unchanged.  I'll submit a PR doing that.

Ultimately I suppose we might want to change the asynchronous logic to actually chain / wait directly on the oneDPL futures, if possible.

Reply:
I just started a PR with this minimal fix #1701. However, `exclusive_scan` and `sort` have the same issue, so those will need to be modified as well.

Reply:
all futures fixed in https://github.com/oneapi-src/oneDPL/pull/1705 and https://github.com/oneapi-src/oneDPL/pull/1701

Reply:
Suggested fix lgtm. For the conditional case, we have to keep in mind there is no guarantee that different overloads return the same future type. It works because the scan with and without initial value boil down to the same pattern and return a future with a type independent of `InitType`. I don't see why that should change in the foreseeable future, worst case the code as is won't compile. Will have to investigate if a guarantee as described makes sense, or a redesign of the future should be considered.

Reply:
> Ultimately I suppose we might want to change the asynchronous logic to actually chain / wait directly on the oneDPL futures, if possible.

The used fan-out fan-in pattern is tricky with the current design because there is no unified future type.
Chaining is possible by design, e.g. [example-of-async-api-usage](https://github.com/oneapi-src/oneDPL/blob/8e929041f3df9ab3714b128405d6b809d3c1c61a/documentation/library_guide/parallel_api/async_api.rst#example-of-async-api-usage).

Reply:
> Suggested fix lgtm. For the conditional case, we have to keep in mind there is no guarantee that different overloads return the same future type. It works because the scan with and without initial value boil down to the same pattern and return a future with a type independent of `InitType`. I don't see why that should change in the foreseeable future, worst case the code as is won't compile. Will have to investigate if a guarantee as described makes sense, or a redesign of the future should be considered.

I suppose we'd need to keep two separate `std::vector`s in that case (which wouldn't be too difficult a fix).  The only other option, I think, would be to use type erasure to encapsulate the DPL futures in some kind of `dpl_generic_future<T>`.